Add EOS S3 GPIO driver

kowalewskijan opened the pull request:
This PR introduces the GPIO driver for EOS S3 SoC and enables the driver for the QuickFeather board.

Note: This PR can be merged after [hal_quicklogic PR](https://github.com/zephyrproject-rtos/hal_quicklogic/pull/2) to avoid conflicts between Zephyr API and HAL.

wtatarski committed:
dts: add EOS S3 GPIO definitions

Add GPIO definitions to QuickLogic EOS S3 devicetree.

Co-authored-by: Jan Kowalewski <jkowalewski@antmicro.com>

Signed-off-by: Wojciech Tatarski <wtatarski@antmicro.com>
Signed-off-by: Jan Kowalewski <jkowalewski@antmicro.com>

wtatarski committed:
drivers: gpio: add EOS S3 GPIO driver

Add GPIO driver for QuickLogic EOS S3 SoC.

Co-authored-by: Jan Kowalewski <jkowalewski@antmicro.com>

Signed-off-by: Wojciech Tatarski <wtatarski@antmicro.com>
Signed-off-by: Jan Kowalewski <jkowalewski@antmicro.com>

wtatarski committed:
boards: quick_feather: enable GPIO driver

Enable GPIO driver for Quick Feather board.

Co-authored-by: Jan Kowalewski <jkowalewski@antmicro.com>

Signed-off-by: Wojciech Tatarski <wtatarski@antmicro.com>
Signed-off-by: Jan Kowalewski <jkowalewski@antmicro.com>

pabigot commented on drivers/gpio/gpio_eos_s3.c:
```diff
@@ -0,0 +1,432 @@
+ * to "port1".
+ */
+static int find_pad(GPIOCfgTypeDef *cfg, int port_num)
+{
```
Can you use the arguments to index a static const array of 16 `PadConfig` instances?  I suspect the cost in flash will be significantly less than the size of this function that constructs them at runtime.

    cfg->xPadConf = pad_array[(8 * port_num) + cfg->ucGpioNum];

pabigot commented on drivers/gpio/gpio_eos_s3.c:
```diff
@@ -0,0 +1,432 @@
+
+#include "gpio_utils.h"
+
+#define MAX_GPIOS_PER_PORT	8
```
If you define this a `8U` you might get away from having to cast values to `(int)` below.

pabigot commented on drivers/gpio/gpio_eos_s3.c:
```diff
@@ -0,0 +1,432 @@
+	}
+
+	/* Determine "port" number */
+	if ((int)(gpio_num_extended / MAX_GPIOS_PER_PORT) == 1) {
```
I don't think the cast here should be necessary, if `MAX_GPIOS_PER_PORT` is unsigned (and, optionally, the RHS is `1U`).

pabigot commented on drivers/gpio/gpio_eos_s3.c:
```diff
@@ -0,0 +1,432 @@
+	uint8_t gpio_num = gpio_num_extended % 8;
+
+	/* Check for an invalid pin configuration */
+	if (((flags & GPIO_INPUT) != 0) && ((flags & GPIO_OUTPUT) != 0)) {
```
`(flags & (GPIO_INPUT | GPIO_OUTPUT)) != 0` ?

Or `(flags & DIR_MASK) != 0` ?

kowalewskijan replied:
I would leave it as is. Checking against e.g. `(flags & DIR_MASK) != 0` will result in "some" value which will correspond to either `GPIO_INPUT` or `GPIO_OUTPUT`. The same check is performed in the mcux driver: https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/gpio/gpio_mcux.c#L48

pabigot replied:
OK, I get it: it doesn't support configuring in both directions.

This possibly should be:

    if (((flags & GPIO_INPUT) != 0) == ((flags & GPIO_OUTPUT) != 0))

because as-is if neither input nor output is specified it'll be configured as an output.

Alternatively just use the switch statement that mcux has after the `if` you point to (which is not necessary, since the switch statement that detects both unsupported configurations.)  Multiple drivers use the switch idiom.

pabigot commented on drivers/gpio/gpio_eos_s3.c:
```diff
@@ -0,0 +1,432 @@
+
+	for (uint8_t gpion = 0; gpion < MAX_GPIOS_PER_PORT; gpion++) {
+		HAL_GPIO_Read(gpion, &pin_value);
+		*value |= (pin_value == 1) ? 1 << gpion : 0;
```
Might be better to have a `gpio_port_value_t mask = 1U` that gets shifted up in the loop body, but if you want to do `1 << x` please use `BIT(x)` from `<sys/utils.h>`.

kowalewskijan replied:
thanks, I switched it to `BIT(x)`

pabigot commented on drivers/gpio/gpio_eos_s3.c:
```diff
@@ -0,0 +1,432 @@
+{
+	ARG_UNUSED(dev);
+
+	for (uint8_t gpion = 0; gpion < MAX_GPIOS_PER_PORT; gpion++) {
```
I'm a bit confused by this.  You only have one device, but above it sounded like you have two ports with eight pins each, which are identified by port*8 + pin.  But this only iterates over 8 pins.

The pin configure code seems to respect the extended GPIO num to select a pad configuration given pin 7 or pin 15, but it seems like this would not do anything for pins on port 1.

I guess it must work....

kowalewskijan replied:
Let me explain. There are total 16 physical pads/pins which can be configured as GPIOs. EOS S3 has only 8 GPIOs, each GPIO has a corresponding pair of physical pins and only one pin can be picked from a pair and used as GPIO. So it is possible to configure only 8 GPIOs connected to 8 pins. Controller needs only information about GPIO number (0-7) which should be read or written, beacuse the corresponding pin is assigned on the configure step.

pabigot commented on dts/bindings/gpio/quicklogic,eos-s3-gpio.yaml:
```diff
@@ -0,0 +1,19 @@
+
+    "#gpio-cells":
+      const: 2
+
```
This needs something like:
```
    ngpios:
      required: true
      const: 16
```
so the common code can determine the appropriate pin mask.  (Also this might help me understand whether the user-facing API provides 8 or 16 pins.)

pabigot reviewed (commented):
Some initial comments.  The architecture of this peripheral is not clear to me.

kowalewskijan commented:
@pabigot I resolved some of your comments and provided a short explanation how the peripheral handles GPIOs.
@pabigot @galak Could you remove the `stale` label from this PR?
Thanks.

pabigot commented:
While I'm looking at this could somebody please do something about #29283?  I have no idea what this board is, and neither will anybody else.

pabigot commented:
I'm pulling out my response to https://github.com/zephyrproject-rtos/zephyr/pull/28765#discussion_r545149330 a direct comment because this is a design-level point:

> Let me explain. There are total 16 physical pads/pins which can be configured as GPIOs. EOS S3 has only 8 GPIOs, each GPIO has a corresponding pair of physical pins and only one pin can be picked from a pair and used as GPIO. So it is possible to configure only 8 GPIOs connected to 8 pins. Controller needs only information about GPIO number (0-7) which should be read or written, beacuse the corresponding pin is assigned on the configure step.

OK, so that's a little more clear.

It looks like this driver is mixing pinmux configuration with GPIO configuration, by giving each of the 8 actual GPIO pins two pin indexes, which encode the pin to use in bit 3 with the true GPIO ordinal in bits 0..2.

So you can configure pin 1 and pin 2 and you'd have two GPIOs active, but if you configure pin 1 and pin 9 then you've got one GPIO (GPIO1 on pad 26) working and have quietly broken subsequent use of pin 1 (GPIO1 on pad 9).

Do I have that right?

If so, this seems way too fragile.

Would you expect to have an application switch between pads at runtime?  If not then I think the GPIO controller devicetree binding should have a bitset property that indicates for the 8 gpios whether they should be connected to the primary (0) or secondary (1) pad for that GPIO.  Then the controller `ngpios` should only claim to have eight pins, and all is good.  In the future when Zephyr has a pinmux API we can add support for changing which pad is the connection for a specific GPIO at runtime if that's useful.

If you do expect to need to move GPIOs between pads right now, we need to figure out another solution, because having two pin ordinals that refer to the same GPIO signal is going to confuse the stuffing out of everybody.

kowalewskijan commented:
Link to the TRM just in case: https://www.quicklogic.com/wp-content/uploads/2020/06/QL-S3-Technical-Reference-Manual-Vol-1.pdf#page=154

Generally there are two peripherals in EOS S3 which controls GPIOs: `IOMux` and `Misc` peripherals. `IOMux` is used to setup pins as GPIOs and driving a pin can be done from `Misc`.

> Do I have that right?

Yep that's right. Well it seems that it is possible to switch between pads in runtime, but currently we can go with non-runtime version of the driver to keep it simple. Also it is noted in the TRM that the `IOMux` pad functions must be determined for a system design, so I believe non-runtime version is a way to go.

pabigot reviewed (changes requested):
Turning https://github.com/zephyrproject-rtos/zephyr/pull/28765#issuecomment-747512399 into a real review:

Rework this so the pads associated with the GPIO are determined at build time, rather than encoding a pad selection into the GPIO ordinal.

kowalewskijan commented:
I've added [pin-cfgs property](https://github.com/zephyrproject-rtos/zephyr/pull/28765/files#diff-f3c56a13fb8772a158e1b2811f0d7bdd6641c620a03b5f0c1d00484b368c101eR17-R35) to EOS S3 devicetree. It is now possible to setup either primary or secondary pin for each GPIO in the build time. @pabigot please take a look

mnkp commented on drivers/gpio/gpio_eos_s3.c:
```diff
@@ -0,0 +1,335 @@
+							    dev, BIT(gpio_num));
+				}
+			}
+			INTR_CTRL->GPIO_INTR |= intr_num;
```
I assume we are writing to the register to clear the interrupt. This should be done immediately after the line
```
uint8_t intr_status = INTR_CTRL->GPIO_INTR;
```
before we call `gpio_fire_callbacks` to avoid losing interrupts. We need to assume the `gpio_fire_callbacks` function can take long to execute.

mnkp commented on drivers/gpio/gpio_eos_s3.c:
```diff
@@ -0,0 +1,335 @@
+		    CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,
+		    &gpio_eos_s3_driver_api);
+
+static int gpio_eos_s3_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+	IRQ_CONNECT(DT_INST_IRQN(0),
+		    DT_INST_IRQ(0, priority),
+		    gpio_eos_s3_isr,
+		    DEVICE_DT_INST_GET(0),
+		    0);
+
+	irq_enable(DT_INST_IRQN(0));
+
+	return 0;
+}
```
It's a non written, de-facto standard to place device instance code at the bottom of the driver file. Let's move this function to just below `static const struct gpio_driver_api gpio_eos_s3_driver_api` where there is currently the declaration.

mnkp reviewed (changes requested):
Let's test the driver with `tests/drivers/gpio/gpio_api_1pin/` testcase.

mnkp commented:
The "Add GPIO support for QuickLogic EOS S3 SoCs." commit relies on `GPIO_EOS_S3` symbol defined in "Add GPIO driver for QuickLogic EOS S3 SoC." commit. We need to squash the two commits.

kowalewskijan commented on drivers/gpio/gpio_eos_s3.c:
```diff
@@ -0,0 +1,335 @@
+	/* Check for an invalid pin configuration */
+	if (((flags & GPIO_INPUT) != 0) == ((flags & GPIO_OUTPUT) != 0)) {
+		return -ENOTSUP;
+	}
```
I've removed the condition initially but I'll have to look up if there should be some additional configurations to be done in the driver.

kowalewskijan commented on dts/bindings/gpio/quicklogic,eos-s3-gpio.yaml:
```diff
@@ -0,0 +1,39 @@
+    "#gpio-cells":
+      const: 2
+
+    pin-cfgs:
```
I renamed it to `pin-secondary-config` and reworked this field to a single byte (with bit masking in the driver).

pabigot reviewed (approved):
Thanks, the things I was concerned about have been addressed.

Does this pass the tests in `tests/drivers/gpio?`

kowalewskijan commented:
@pabigot I've just checked and there are still some adjustments that have to be done to make GPIO tests pass on hardware. I'll notify here when I'll have them green.

mnkp commented on drivers/gpio/gpio_eos_s3.c:
```diff
@@ -0,0 +1,311 @@
+			return -EINVAL;
+		}
+
+		data->gpio_irq_num[gpio_cfg.ucGpioNum] = BIT((uint8_t)irq_num);
+		/* Disable IRQ */
+		INTR_CTRL->GPIO_INTR_EN_M4 &= ~data->gpio_irq_num[gpio_num];
```
You use `gpio_cfg.ucGpioNum` and `gpio_num` interchangeably. Let's maybe stick to `gpio_num`, the code will be easier to read.

kowalewskijan replied:
refactored, thanks

mnkp commented on drivers/gpio/gpio_eos_s3.c:
```diff
@@ -0,0 +1,311 @@
+			return -EINVAL;
+		}
+
+		data->gpio_irq_num[gpio_cfg.ucGpioNum] = BIT((uint8_t)irq_num);
```
`gpio_cfg.ucGpioNum` -> `gpio_num`

mnkp commented on drivers/gpio/gpio_eos_s3.c:
```diff
@@ -0,0 +1,311 @@
+	PadConfig pad_conf;
+
+	gpio_cfg.ucGpioNum = gpio_num;
+	pad_conf = pad_configs[(8 * is_secondary) + gpio_num];
```
Similar code is in `gpio_eos_s3_configure`. It might be worth to turn it into a small function taking `dev` and `gpio_num` as parameter and returning PadConfig. `config` and `is_secondary` would be gone from the variable definition block.

kowalewskijan replied:
extracted the common part into [gpio_eos_s3_pad_select](https://github.com/zephyrproject-rtos/zephyr/pull/28765/files#diff-46c5dbff8c5798f8c66fc441672c6fb33fbe7ce2f843e9715e464ce1b97e068eR71-R78) function

Commit:
tests: drivers: gpio: provide quickfeather overlay

This commit provides the necessary overlay to make the test pass
on the QuickFeather hardware.

Signed-off-by: Jan Kowalewski <jkowalewski@antmicro.com>

mnkp commented on drivers/gpio/gpio_eos_s3.c:
```diff
@@ -0,0 +1,325 @@
+	{.ucPin = PAD_45, .ucFunc = PAD45_FUNC_SEL_GPIO_7},
+};
+
+static PadConfig gpio_eos_s3_pad_select(const struct device *dev,
```
Sorry, I wasn't very precise in my last request. It's a minor improvement since PadConfig is small, however in the code 
```
gpio_cfg.xPadConf = &pad_config;
```
we need the address of the PadConfig and not the structure itself. We could return here a pointer to PadConfig instead.

mnkp replied:
I have warned about const-ness but didn't notice there is no way around it. What I have asked for was a bad idea. You removed const from
```
static const PadConfig pad_configs[] = {
```
and this moved the array to RAM. `PadConfig` is actually not that small, sizeof(PadConfig) is 16. So the the whole array takes 256 bytes. That's a lot of RAM. Even for flash it's not negligible.

Let's either go to the previous design or we could optimize this code. All `PAD*_FUNC_SEL_GPIO_*` have identical value. It's unlikely that in the future this will change. If you add a compile time assert using `BUILD_ASSERT` that would verify that all `PAD*_FUNC_SEL_GPIO_*` are equal then you could
```
#define PAD_FUNC_SEL_GPIO PAD6_FUNC_SEL_GPIO_0
```
and use `PAD_FUNC_SEL_GPIO` in the code.

Then we would only need an array to store relation between gpio_num and pad number.

This `BUILD_ASSERT` is going to look ugly. If you don't like it let's go to the previous design. Sorry about all that.

kowalewskijan replied:
No problem, I reverted changes and went back to the previous implementation.

mnkp commented on drivers/gpio/gpio_eos_s3.c:
```diff
@@ -0,0 +1,325 @@
+	struct gpio_eos_s3_data *data = dev->data;
+	uint8_t intr_status = INTR_CTRL->GPIO_INTR & GPIOS_MASK;
+	
+	if (!intr_status) {
+		return;
+	}
```
I'm not familiar with the architecture and I may be wrong but that doesn't look right. If you entered the `gpio_eos_s3_isr` it means there was some interrupt that was active. It's not an interrupt that we find interesting since it's outside of `GPIOS_MASK`, however it is there. You didn't clear it. Soon after you exit the interrupt routine you'll enter it again, the interrupt is still marked as active. In principle your program will freeze here in the endless loop.

On architectures that do not automatically clear interrupt status register on read operation - and it looks like this is the one - it's best to do it manually just after you read the status register. It also ensures that if there are two interrupt events close to each other and the interrupt routine is long both interrupts will be noticed. If you clear status register late you may miss an interrupt.

mnkp reviewed (commented):
LGTM, though I still have some comments. If you address the comments and let us know that the testcases are passing I will be happy to approve this PR.

kowalewskijan commented:
@pabigot @mnkp I pushed some changes regarding handling interrupts which was the main problem to pass GPIO tests on hardware. Currently with the recent changes all tests are green on the device.

Additionally, I also updated the QuickFeather doc to point that the GPIO feature is supported.

mnkp commented on drivers/gpio/gpio_eos_s3.c:
```diff
@@ -0,0 +1,403 @@
+		data->gpio_irq_mask |= BIT((uint32_t)irq_num);
+
+		/* Enable IRQ */
+		INTR_CTRL->GPIO_INTR_EN_M4 |= BIT((uint32_t)irq_num);
```
I didn't find detailed information about interrupt controller in the datasheet, however we probably should clear the interrupt status register before enabling the interrupt. Something like
```
	INTR_CTRL->GPIO_INTR |= BIT((uint32_t)irq_num);
```
On many architectures if the status register was dirty and we enable the interrupt it will trigger immediately.

kowalewskijan replied:
I agree, this is something I missed and it is definitely worth adding - fixed.

mnkp commented on drivers/gpio/gpio_eos_s3.c:
```diff
@@ -0,0 +1,403 @@
+	uint32_t intr_status = INTR_CTRL->GPIO_INTR;
+	uint32_t enabled_int = intr_status & data->gpio_irq_mask;
+
+	INTR_CTRL->GPIO_INTR |= enabled_int;
```
Again, I didn't find detailed information about interrupt controller in the datasheet. It may work differently than I assume. However, it's an error for the interrupt routine not to clear all of the active interrupt sources. If we don't clear the status bit we will enter the interrupt handler as soon as we exit it. Effectively it works as an endless loop. I propose to write it as
```
	INTR_CTRL->GPIO_INTR |= intr_status;
```
You may add an `__ASSERT` to verify that there are no bugs in the driver code and there was no other interrupt than that allowed by `data->gpio_irq_mask`.

kowalewskijan replied:
We should clear the interrupt statuses as you suggested. In the result I dropped the `gpio_irq_mask` because the control of fired callbacks is done already with the implemented condition: `data->gpio_irqs[irq_num] != DISABLED_GPIO_IRQ`.

I've tested the changes and all tests passed.

mnkp reviewed (commented):
I propose to squash all the "boards: arm: quick_feather:" commits into one, or at least the following two
- "boards: quick_feather: enable GPIO driver"
- "boards: arm: quick_feather: add GPIO feature to docs"

kowalewskijan commented:
> I propose to squash all the "boards: arm: quick_feather:" commits into one, or at least the following two
> 
> * "boards: quick_feather: enable GPIO driver"
> * "boards: arm: quick_feather: add GPIO feature to docs"

@mnkp squashed all the `boards` commits
